File: 2024/4/main.pt1.cpp
#include<stdio.h>
#include<iostream>

using namespace std;

class XmasSolver
{
    private:
        int n, m;
        char arr[150][150];
        const string MAGIC_WORD = "XMAS";
    public:
        int find_rest(int, int, int, int, int);
        int find_x();

    XmasSolver(int n, int m, char arr[150][150])
    {
        this->n = n;
        this->m = m;

        for(int i = 0; i < n ; i++)
        {
            for(int j = 0; j < m ; j++)
            {
                this->arr[i][j] = arr[i][j];
            }
        }
    }
};

int XmasSolver::find_rest(int i, int j, int dir_x, int dir_y, int curr_letter)
{
    if(curr_letter == this->MAGIC_WORD.size())
        return 1;

    i += dir_x;
    j += dir_y;

    if(i < this->n && i >= 0 && j < this->m && j >= 0)
    {    
        if(this->arr[i][j] == this->MAGIC_WORD[curr_letter])
        {
            return find_rest(i, j, dir_x, dir_y, curr_letter + 1);
        }
    }
    
    return 0;
}

int XmasSolver::find_x()
{
    int ans = 0;
    const int DIR_X[] = {0, 0, 1, 1, 1, -1, -1, -1};
    const int DIR_Y[] = {-1, 1, -1, 0, 1, -1, 0, 1};

    for(int i = 0 ; i < this->n ; i++)
    {
        for(int j = 0 ; j < this->m ; j++)
        {
            if(this->arr[i][j] == 'X')
            {
                for(int d = 0; d < 8; d++)
                {
                    ans += find_rest(i, j, DIR_X[d], DIR_Y[d], 1);
                }
            }
        }
    }

    return ans;
}

int main()
{
    FILE *f = fopen("input.txt", "r");
    char c, arr[150][150];
    int i = 0, j = 0;

    while(fgets(arr[i], 150, f) != NULL)
    {
        i++;
    }

    for(j = 0 ; arr[0][j] != '\0' ; j++)
    {}

    XmasSolver solver = XmasSolver(i, j, arr);
    int ans = solver.find_x();

    printf("%d", ans);

    return 0;
}

File: 2024/12/main.pt1.py

class Solver:
    def __init__(self, n, m, arr):
        self.n = n
        self.m = m

        self.DIR_X = [1, -1, 0, 0]
        self.DIR_Y = [0, 0, 1, -1]
        
        self.visited = [[False for j in range(m)] for i in range(n)]
        self.arr = arr

    def correct_indices(self, i, j):
        return i >= 0 and i < self.n and j >= 0 and j < self.m

    def solve(self, i, j, area, perimeter, curr_letter):
        if (self.visited[i][j] == True):
            return area, perimeter
        
        self.visited[i][j] = True
        area += 1

        for d in range(4):
            new_i = i + self.DIR_X[d]
            new_j = j + self.DIR_Y[d]

            if (not self.correct_indices(new_i, new_j)):
                perimeter += 1

            if (self.correct_indices(new_i, new_j)):
                if (self.arr[new_i][new_j] == curr_letter):
                    area, perimeter = self.solve(new_i, new_j, area, perimeter, curr_letter)
                elif (self.arr[new_i][new_j] != curr_letter):
                    perimeter += 1

        return area, perimeter
    
    def init_solve(self):
        ans = 0

        for i in range(self.n):
            for j in range(self.m):
                if (self.visited[i][j] == False):
                    a, p = self.solve(i, j, 0, 0, self.arr[i][j])
                    ans += a * p

        return ans

def main():
    
    f = open("./input.txt", "r")

    ans = 0

    garden = []
    n = None
    m = None

    for line in f:
        l = list(map(lambda x: x, line.strip()))
        garden.append(l)

        if m == None:
            m = len(l)

    if n == None:
        n = len(garden)

    print(n, m)
    s = Solver(n, m, garden)
    ans = s.init_solve()

    print(ans)

if __name__ == "__main__":
    main()